File: app/src/main/java/it/niedermann/nextcloud/deck/ui/exception/ExceptionHandler.java
package it.niedermann.nextcloud.deck.ui.exception;

import android.app.Activity;

import androidx.annotation.NonNull;

import it.niedermann.nextcloud.deck.DeckLog;

public class ExceptionHandler implements Thread.UncaughtExceptionHandler {

    @NonNull
    private final Activity activity;

    public ExceptionHandler(@NonNull Activity activity) {
        this.activity = activity;
    }

    @Override
    public void uncaughtException(@NonNull Thread t, @NonNull Throwable e) {
        DeckLog.logError(e);
        activity.getApplicationContext().startActivity(ExceptionActivity.createIntent(activity, e));
        activity.finish();
        Runtime.getRuntime().exit(0);
    }
}

File: app/android/src/main/java/it/niedermann/nextcloud/deck/ui/exception/ExceptionHandler.kt
package it.niedermann.nextcloud.deck.ui.exception

import android.content.Context
import android.util.Log

class ExceptionHandler(
    private val context: Context
) : Thread.UncaughtExceptionHandler {

    override fun uncaughtException(t: Thread, e: Throwable) {
        try {
            Log.e("ExceptionHandler", "Uncaught exception in thread ${t.name}", e)
            val intent = ExceptionActivity.createIntent(context, e)
            context.startActivity(intent)
        } catch (error: Exception) {
            Log.e("ExceptionHandler", "Error in ExceptionHandler", error)
        } finally {
            Runtime.getRuntime().exit(1)
        }
    }

    companion object {
        fun initialize(context: Context) {
            val currentHandler = Thread.getDefaultUncaughtExceptionHandler()
            if (currentHandler is ExceptionHandler) return
            Thread.setDefaultUncaughtExceptionHandler(ExceptionHandler(context))
        }
    }
}
